File: tasks/bench-eager-forms/fixtures/reference/eager.py
import sys, re
def tok(s): return s.replace('(', ' ( ').replace(')', ' ) ').split()
def parse(t, i):
    if t[i] == '(':
        i += 1; items = []
        while t[i] != ')':
            e, i = parse(t, i); items.append(e)
        return ('form', items), i + 1
    if re.fullmatch(r'-?\d+', t[i]): return ('int', int(t[i])), i + 1
    return ('name', t[i]), i + 1
def run(src):
    out, funcs, glob, counters = [], {}, {}, [0] * 10
    depth = [0]
    def ev(e, env):
        k = e[0]
        if k == 'int': return e[1]
        if k == 'name': return env[e[1]] if e[1] in env else glob[e[1]]
        op = e[1][0][1]
        vals = [ev(a, env) for a in e[1][1:]]
        if op == '+': return vals[0] + vals[1]
        if op == '-': return vals[0] - vals[1]
        if op == '*': return vals[0] * vals[1]
        if op == '<': return int(vals[0] < vals[1])
        if op == '=': return int(vals[0] == vals[1])
        if op == 'and': return int(vals[0] != 0 and vals[1] != 0)
        if op == 'or': return int(vals[0] != 0 or vals[1] != 0)
        if op == 'not': return int(vals[0] == 0)
        if op == 'if': return vals[1] if vals[0] != 0 else vals[2]
        if op == 'seq': return vals[1]
        if op == 'out': out.append(vals[0]); return vals[0]
        if op == 'bump': counters[vals[0]] += 1; return counters[vals[0]]
        if op == 'peek': return counters[vals[0]]
        ps, body = funcs[op]
        if depth[0] + 1 > 5: return 0
        depth[0] += 1
        try: return ev(body, dict(zip(ps, vals)))
        finally: depth[0] -= 1
    for line in src.split('\n'):
        line = line.strip()
        if not line or line.startswith('#'): continue
        m = re.match(r'^(LET|DEF|SHOW)\s+(.*)$', line)
        head, rest = m.group(1), m.group(2)
        if head == 'SHOW':
            e, _ = parse(tok(rest), 0); out.append(ev(e, {}))
        elif head == 'LET':
            n, r = rest.split('=', 1); e, _ = parse(tok(r), 0); glob[n.strip()] = ev(e, {})
        else:
            l, r = rest.split('=', 1); t = l.split(); e, _ = parse(tok(r), 0); funcs[t[0]] = (t[1:], e)
    return out
if __name__ == '__main__':
    print(' '.join(str(x) for x in run(open(sys.argv[1]).read())))

File: tasks/bench-eager-forms/fixtures/reference/eager.mjs
// Independent second implementation: compile to closures, then run.
import { readFileSync } from 'node:fs';
export function runEager(src) {
  const out = [], funcs = new Map(), glob = new Map(), cnt = Array(10).fill(0);
  let depth = 0;
  const tokens = (s) => s.replace(/\(/g, ' ( ').replace(/\)/g, ' ) ').trim().split(/\s+/);
  function parseForm(t, pos) {
    const x = t[pos.i];
    if (x !== '(') { pos.i++; if (/^-?\d+$/.test(x)) return { k: 'int', v: BigInt(x) }; return { k: 'name', n: x }; }
    pos.i++; const op = t[pos.i++]; const args = [];
    while (t[pos.i] !== ')') args.push(parseForm(t, pos));
    pos.i++; return { k: 'form', op, args };
  }
  const b = (x) => (x ? 1n : 0n);
  function ev(e, env) {
    if (e.k === 'int') return e.v;
    if (e.k === 'name') return env.has(e.n) ? env.get(e.n) : glob.get(e.n);
    const vs = []; for (const a of e.args) vs.push(ev(a, env));
    switch (e.op) {
      case '+': return vs[0] + vs[1]; case '-': return vs[0] - vs[1]; case '*': return vs[0] * vs[1];
      case '<': return b(vs[0] < vs[1]); case '=': return b(vs[0] === vs[1]);
      case 'and': return b(vs[0] !== 0n && vs[1] !== 0n); case 'or': return b(vs[0] !== 0n || vs[1] !== 0n);
      case 'not': return b(vs[0] === 0n); case 'if': return vs[0] !== 0n ? vs[1] : vs[2]; case 'seq': return vs[1];
      case 'out': out.push(vs[0]); return vs[0];
      case 'bump': { const k = Number(vs[0]); cnt[k]++; return BigInt(cnt[k]); }
      case 'peek': return BigInt(cnt[Number(vs[0])]);
    }
    const f = funcs.get(e.op);
    if (depth >= 5) return 0n;
    const local = new Map(f.ps.map((p, i) => [p, vs[i]]));
    depth++; const r = ev(f.body, local); depth--; return r;
  }
  for (const raw of src.split('\n')) {
    const line = raw.trim(); if (!line || line.startsWith('#')) continue;
    const sp = line.indexOf(' '); const head = line.slice(0, sp); const rest = line.slice(sp + 1);
    if (head === 'SHOW') out.push(ev(parseForm(tokens(rest), { i: 0 }), new Map()));
    else if (head === 'LET') { const q = rest.indexOf('='); glob.set(rest.slice(0, q).trim(), ev(parseForm(tokens(rest.slice(q + 1)), { i: 0 }), new Map())); }
    else { const q = rest.indexOf('='); const l = rest.slice(0, q).trim().split(/\s+/); funcs.set(l[0], { ps: l.slice(1), body: parseForm(tokens(rest.slice(q + 1)), { i: 0 }) }); }
  }
  return out.map(String);
}
if (process.argv[2]) console.log(runEager(readFileSync(process.argv[2], 'utf8')).join(' '));
